engine/classes/Elgg/Notifications/NotificationsService.php

```diff
@@ -197,129 +215,304 @@ public function processQueue($stopTime) {
+					if ($enabled) {
+						$methods[] = $method;
+					}
+				}
+			}
+
+			$subscriptions[$recipient->guid] = $methods;
 		}
 
-		if (($object instanceof ElggEntity) && !has_access_to_entity($object, $recipient)) {
-			return false;
+		$hook_params = [
+			'event' => $params['event'],
+			'origin' => $params['origin'],
+			'methods_override' => $methods_override,
+		];
+		$subscriptions = $this->hooks->trigger('get', 'subscriptions', $hook_params, $subscriptions);
+		
+		$params['subscriptions'] = $subscriptions;
+
+		// return false to stop the default notification sender
+		if ($this->hooks->trigger('send:before', 'notifications', $params, true)) {
+			$deliveries = $this->sendNotifications($event, $subscriptions, $params);
 		}
+		$params['deliveries'] = $deliveries;
+		$this->hooks->trigger('send:after', 'notifications', $params);
 
-		$language = $recipient->language;
-		$params = array(
-			'event' => $event,
-			'method' => $method,
-			'recipient' => $recipient,
-			'language' => $language,
-			'object' => $object,
-		);
+		return $deliveries;
+	}
+
+	/**
+	 * Send a notification to a subscriber
+	 *
+	 * @param NotificationEvent $event  The notification event
+	 * @param int               $guid   The guid of the subscriber
+	 * @param string            $method The notification method
+	 * @param array             $params Default notification params
+	 * @return bool
+	 * @access private
+	 */
+	protected function sendNotification(NotificationEvent $event, $guid, $method, array $params = []) {
 
-		$subject = $this->getNotificationSubject($event, $recipient);
-		$body = $this->getNotificationBody($event, $recipient);
+		$actor = $event->getActor();
+		$object = $event->getObject();
 
-		$notification = new \Elgg\Notifications\Notification($event->getActor(), $recipient, $language, $subject, $body, '', $params);
+		if ($event instanceof InstantNotificationEvent) {
+			$recipient = $this->entities->get($guid);
+			/* @var \ElggEntity $recipient */
+			$subject = elgg_extract('subject', $params, '');
+			$body = elgg_extract('body', $params, '');
+			$summary = elgg_extract('summary', $params, '');
+		} else {
+			$recipient = $this->entities->get($guid, 'user');
+			/* @var \ElggUser $recipient */
+			if (!$recipient || $recipient->isBanned()) {
+				return false;
+			}
+		
+			// don't notify the creator of the content
+			if ($recipient->getGUID() == $event->getActorGUID()) {
+				return false;
+			}
+			
+			if (!$actor || !$object) {
+				return false;
+			}
+
+			if ($object instanceof ElggEntity) {
+				if (!in_array($object->access_id, [ACCESS_PUBLIC, ACCESS_LOGGED_IN]) && !has_access_to_entity($object, $recipient)) {
+					// In this context we want to make sure public and logged in access is considered accessible
+					// in spite of what access SQL may say: when running in cron we don't have a logged in user,
+					// in walled garden public access may not be in user's access list
```

Comment:
For some reason tests were failing on this. I will need to investigate further. Basically when we don't have a logged in user, access ids 1 and 2 are not in access array hence get_entity fails